File: tests/ts/test_clean_for_speech.ts
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import { cleanForSpeech } from "../../plugin/voice.ts";

describe("cleanForSpeech", () => {

  // --- Fenced code blocks ---

  it("removes triple-backtick fenced code blocks", () => {
    const input = "Here is code:\n```python\nx = 1\n```\nDone.";
    assert.equal(cleanForSpeech(input), "Here is code: Done.");
  });

  it("removes tilde fenced code blocks", () => {
    const input = "Code:\n~~~js\nconst y = 2;\n~~~\nEnd.";
    assert.equal(cleanForSpeech(input), "Code: End.");
  });

  it("removes multi-line fenced code blocks", () => {
    const input = "Start\n```\nline1\nline2\nline3\n```\nFinish";
    assert.equal(cleanForSpeech(input), "Start Finish");
  });

  // --- Inline code ---

  it("preserves inline code text but removes backticks", () => {
    assert.equal(cleanForSpeech("Use `large-v3` for STT"), "Use large-v3 for STT");
  });

  it("handles multiple inline code spans", () => {
    assert.equal(
      cleanForSpeech("Set `foo` and `bar` then run"),
      "Set foo and bar then run",
    );
  });

  // --- Bold / italic / strikethrough ---

  it("removes bold markers preserving text", () => {
    assert.equal(cleanForSpeech("**Important:** tests passed"), "Important: tests passed");
  });

  it("removes strikethrough markers preserving text", () => {
    assert.equal(cleanForSpeech("~~old text~~ new text"), "old text new text");
  });

  it("removes single-asterisk italic markers", () => {
    assert.equal(cleanForSpeech("This is *italic* text"), "This is italic text");
  });

  it("removes underscore emphasis", () => {
    assert.equal(cleanForSpeech("file_test and _important_ value"), "file_test and important value");
  });

  // --- Links ---

  it("converts markdown links to visible text only", () => {
    assert.equal(
      cleanForSpeech("See [OpenCode](https://opencode.ai) docs"),
      "See OpenCode docs",
    );
  });

  it("handles multiple links in one string", () => {
    assert.equal(
      cleanForSpeech("[a](http://a.com) and [b](http://b.com)"),
      "a and b",
    );
  });

  // --- URLs ---

  it("removes bare HTTP URLs", () => {
    assert.equal(
      cleanForSpeech("Visit https://example.com today"),
      "Visit today",
    );
  });

  it("removes bare HTTPS URLs", () => {
    assert.equal(
      cleanForSpeech("Go to http://foo.bar/x?y=1 now"),
      "Go to now",
    );
  });

  // --- Paths ---

  it("shortens Unix paths to final component", () => {
    assert.equal(
      cleanForSpeech("Edit /home/user/project/src/voice.py please"),
      "Edit voice.py please",
    );
  });

  it("shortens deep paths to final directory or file", () => {
    assert.equal(
      cleanForSpeech("Changed /home/o/git/111-AI/opencode-voice-kokoro/scripts/voice.py"),
      "Changed voice.py",
    );
  });

  // --- Headings ---

  it("removes heading markers", () => {
    assert.equal(cleanForSpeech("### Installation"), "Installation");
  });

  it("removes atx heading with text after", () => {
    assert.equal(
      cleanForSpeech("## Title\nSome content"),
      "Title Some content",
    );
  });

  // --- Blockquotes ---

  it("removes blockquote markers", () => {
    assert.equal(cleanForSpeech("> This is a quote"), "This is a quote");
  });

  // --- Lists ---

  it("removes unordered list markers", () => {
    assert.equal(cleanForSpeech("- item one\n- item two"), "item one item two");
  });

  it("removes ordered list numbering", () => {
    assert.equal(cleanForSpeech("1. First\n2. Second"), "First Second");
  });

  it("removes asterisk list markers", () => {
    assert.equal(cleanForSpeech("* item A\n* item B"), "item A item B");
  });

  // --- Horizontal rules ---

  it("removes horizontal rules", () => {
    assert.equal(cleanForSpeech("Above\n---\nBelow"), "Above Below");
  });

  // --- Whitespace ---

  it("collapses multiple spaces into one", () => {
    assert.equal(cleanForSpeech("too    many     spaces"), "too many spaces");
  });

  it("collapses newlines into spaces", () => {
    assert.equal(cleanForSpeech("line one\nline two\nline three"), "line one line two line three");
  });

  it("trims leading and trailing whitespace", () => {
    assert.equal(cleanForSpeech("  hello world  "), "hello world");
  });

  // --- Edge cases ---

  it("handles empty string", () => {
    assert.equal(cleanForSpeech(""), "");
  });

  it("handles whitespace-only string", () => {
    assert.equal(cleanForSpeech("   \n\t  \n  "), "");
  });

  it("handles plain text with no markdown", () => {
    assert.equal(cleanForSpeech("The tests passed successfully."), "The tests passed successfully.");
  });

  // --- Combined real-world OpenCode output ---

  it("cleans a realistic OpenCode response", () => {
    const input = `I updated the server configuration.

\`\`\`python
model = WhisperModel("large-v3", device="cuda")
\`\`\`

The tests now pass. Changed **src/asr.py** and \`src/config.py\`.

See [docs](https://example.com/docs) for details.

### Summary
- Fixed the import
- Updated config at /home/user/project/config.yaml

> Note: restart the service.`;

    const result = cleanForSpeech(input);
    // Should not contain code blocks, backticks, bold markers, URLs, paths, headings, list markers, or blockquote markers
    assert.ok(!result.includes("```"));
    assert.ok(!result.includes("WhisperModel"));
    assert.ok(!result.includes("**"));
    assert.ok(!result.includes("`"));
    assert.ok(!result.includes("https://"));
    assert.ok(!result.includes("/home/"));
    assert.ok(!result.includes("###"));
    assert.ok(!result.includes("- Fixed"));
    assert.ok(!result.includes("> Note"));
    // Should contain the natural-language sentences
    assert.ok(result.includes("I updated the server configuration."));
    assert.ok(result.includes("The tests now pass."));
    assert.ok(result.includes("Fixed the import"));
    assert.ok(result.includes("restart the service."));
  });
});

File: tests/test_clean_for_speech.py
"""Tests for scripts/clean_for_speech.py — parity with tests/ts/test_clean_for_speech.ts."""

from __future__ import annotations

import importlib.util
import sys
from pathlib import Path

import pytest

ROOT = Path(__file__).resolve().parents[1]
MODULE_PATH = ROOT / "scripts" / "clean_for_speech.py"


def _load_clean():
    spec = importlib.util.spec_from_file_location("clean_for_speech", MODULE_PATH)
    assert spec and spec.loader
    mod = importlib.util.module_from_spec(spec)
    sys.modules["clean_for_speech"] = mod
    spec.loader.exec_module(mod)
    return mod.clean_for_speech


clean_for_speech = _load_clean()


class TestFencedCodeBlocks:
    def test_removes_triple_backtick(self):
        assert clean_for_speech("Here is code:\n```python\nx = 1\n```\nDone.") == (
            "Here is code: Done."
        )

    def test_removes_tilde_fenced(self):
        assert clean_for_speech("Code:\n~~~js\nconst y = 2;\n~~~\nEnd.") == "Code: End."

    def test_removes_multi_line_fenced(self):
        assert clean_for_speech("Start\n```\nline1\nline2\nline3\n```\nFinish") == (
            "Start Finish"
        )


class TestInlineCode:
    def test_preserves_inline_code_text(self):
        assert clean_for_speech("Use `large-v3` for STT") == "Use large-v3 for STT"

    def test_multiple_inline_spans(self):
        assert clean_for_speech("Set `foo` and `bar` then run") == (
            "Set foo and bar then run"
        )


class TestEmphasis:
    def test_bold(self):
        assert clean_for_speech("**Important:** tests passed") == (
            "Important: tests passed"
        )

    def test_strikethrough(self):
        assert clean_for_speech("~~old text~~ new text") == "old text new text"

    def test_italic_asterisk(self):
        assert clean_for_speech("This is *italic* text") == "This is italic text"

    def test_underscore_emphasis(self):
        assert clean_for_speech("file_test and _important_ value") == (
            "file_test and important value"
        )


class TestLinksAndUrls:
    def test_markdown_link(self):
        assert clean_for_speech("See [OpenCode](https://opencode.ai) docs") == (
            "See OpenCode docs"
        )

    def test_multiple_links(self):
        assert clean_for_speech("[a](http://a.com) and [b](http://b.com)") == "a and b"

    def test_bare_https(self):
        assert clean_for_speech("Visit https://example.com today") == "Visit today"

    def test_bare_http(self):
        assert clean_for_speech("Go to http://foo.bar/x?y=1 now") == "Go to now"


class TestPaths:
    def test_shortens_unix_path(self):
        assert clean_for_speech("Edit /home/user/project/src/voice.py please") == (
            "Edit voice.py please"
        )

    def test_shortens_deep_path(self):
        assert clean_for_speech(
            "Changed /home/o/git/111-AI/opencode-voice-kokoro/scripts/voice.py"
        ) == "Changed voice.py"


class TestStructure:
    def test_heading(self):
        assert clean_for_speech("### Installation") == "Installation"

    def test_heading_with_body(self):
        assert clean_for_speech("## Title\nSome content") == "Title Some content"

    def test_blockquote(self):
        assert clean_for_speech("> This is a quote") == "This is a quote"

    def test_unordered_list(self):
        assert clean_for_speech("- item one\n- item two") == "item one item two"

    def test_ordered_list(self):
        assert clean_for_speech("1. First\n2. Second") == "First Second"

    def test_asterisk_list(self):
        assert clean_for_speech("* item A\n* item B") == "item A item B"

    def test_horizontal_rule(self):
        assert clean_for_speech("Above\n---\nBelow") == "Above Below"


class TestWhitespaceAndEdges:
    def test_collapse_spaces(self):
        assert clean_for_speech("too    many     spaces") == "too many spaces"

    def test_collapse_newlines(self):
        assert clean_for_speech("line one\nline two\nline three") == (
            "line one line two line three"
        )

    def test_trim(self):
        assert clean_for_speech("  hello world  ") == "hello world"

    def test_empty(self):
        assert clean_for_speech("") == ""

    def test_whitespace_only(self):
        assert clean_for_speech("   \n\t  \n  ") == ""

    def test_plain_text(self):
        assert clean_for_speech("The tests passed successfully.") == (
            "The tests passed successfully."
        )


class TestRealistic:
    def test_realistic_agent_response(self):
        input_text = """I updated the server configuration.

```python
model = WhisperModel("large-v3", device="cuda")
```

The tests now pass. Changed **src/asr.py** and `src/config.py`.

See [docs](https://example.com/docs) for details.

### Summary
- Fixed the import
- Updated config at /home/user/project/config.yaml

> Note: restart the service."""

        result = clean_for_speech(input_text)
        assert "```" not in result
        assert "WhisperModel" not in result
        assert "**" not in result
        assert "`" not in result
        assert "https://" not in result
        assert "/home/" not in result
        assert "###" not in result
        assert "- Fixed" not in result
        assert "> Note" not in result
        assert "I updated the server configuration." in result
        assert "The tests now pass." in result
        assert "Fixed the import" in result
        assert "restart the service." in result
